nanofi/src/core/file_utils.c

```diff
@@ -0,0 +1,54 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "api/nanofi.h"
+#include "core/string_utils.h"
+#include "core/file_utils.h"
+
+token_list tail_file(const char * file_path, char delim, int curr_offset) {
+    token_list tkn_list;
+    memset(&tkn_list, 0, sizeof(struct token_list));
+
+    if (!file_path) {
+        return tkn_list;
+    }
+
+    int max_bytes_read = 4096;
+    char buff[max_bytes_read + 1];
+    memset(buff,'\0', max_bytes_read);
+    FILE * fp = fopen(file_path, "rb");
+    if (!fp) return tkn_list;
+    fseek(fp, curr_offset, SEEK_SET);
+
+    int bytes_read = 0;
+    while ((bytes_read = fread(buff, 1, max_bytes_read, fp)) > 0) {
```

Comment:
I think this modification preserved and amplified the original problem with the logic of this processor.
TailFile, in my interpretation, should do the following on each on_trigger:
 * Restore its previous state (its offset in the file)
 * Repeat until it reaches EOF (or an error occurs):
   * Read the file until a delimiter is reached.
   * Create a FlowFile from the token.
 * Store the position in the file after the end of the last token from which a FlowFile was created

Unless I am mistaken, what this implementation does is:
 * Repeat until it reaches EOF (or an error occurs):
   * Read at most 4096 bytes from the file
   * Find all completely delimited tokens in the resulting buffer
   * Discard the last part of the buffer (or all of it, if there was no delimiter in the read bytes)

This means that every buffer will be skipped that does not contain a delimiter, and the last non-delimited parts of every buffer will be also skipped.
The last version of this PR introduces an additional bug: because the total_bytes contains only the non-skipped parts of the buffers, the resulting total_bytes, which will be later used for determining the offset in the file will be practically a random offset in the file.

Please review and fix the business logic and write tests, including tests for long (>> buffer) lines, delimiters occuring at buffer-boundary and delimiters occuring at non buffer-boundary positions.

Reply:
I haven't fully reviewed this but I think some of this can be a follow on discussion. Calling this "tail file" is a bit of a misnomer in regards to what we're attempting to achieve. There will be slight differences -- and those should be captured a little further in may be some follow on discussion. I do think that the current logic if laid out as such is potentially a bug -- but I envision slight changes in how this functionality is implemented/used